service/controller.go

```diff
@@ -1645,7 +1659,80 @@ func addMetaData(params map[string]string) map[string]string {
 	}
 	return headerMetadata
 }
-func (s *service) ControllerGetVolume(context.Context,
-	*csi.ControllerGetVolumeRequest) (*csi.ControllerGetVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+func (s *service) ControllerGetVolume(ctx context.Context,
+	req *csi.ControllerGetVolumeRequest) (*csi.ControllerGetVolumeResponse, error) {
+
+	// Fetch log handler
+	ctx, log, runID := GetRunIDLog(ctx)
+
+	volID := req.GetVolumeId()
+	if volID == "" {
+		return nil, status.Error(codes.FailedPrecondition, utils.GetMessageWithRunID(runID, "no VolumeID found in request"))
+	}
+	volName, exportID, accessZone, clusterName, err := utils.ParseNormalizedVolumeID(ctx, volID)
+
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, utils.GetMessageWithRunID(runID, err.Error()))
+	}
+
+	ctx, log = setClusterContext(ctx, clusterName)
+	log.Debugf("Cluster Name: %v", clusterName)
+
+	isiConfig, err := s.getIsilonConfig(ctx, &clusterName)
+	isiPath := isiConfig.IsiPath
+
+	if !isiConfig.isiSvc.IsVolumeExistent(ctx, isiPath, volName, "") {
+		return nil, status.Error(codes.InvalidArgument, utils.GetMessageWithRunID(runID, "volume does not exists at this path %v", isiPath))
+	}
+
+	//Fetch volume details
+	volume, err := isiConfig.isiSvc.GetVolume(ctx, isiPath, "", volName)
+	if err != nil {
+		log.Errorf("error in getting '%s' volume '%v'", volName, err)
+		return nil, err
+	}
+
+	//Fetch export clients list
+	exports, err := isiConfig.isiSvc.GetExportByIDWithZone(ctx, exportID, accessZone)
+	if err != nil {
+		jsonError, ok := err.(*isiApi.JSONError)
+		if ok {
+			if jsonError.StatusCode != 404 {
+				return nil, err
+			}
+		}
+		return nil, err
+	}
+
+	abnormal, message := s.getVolumeCondition(ctx, isiConfig)
+	//remove localhost from the clients
+	exportList := removeString(*exports.Clients, "localhost")
+	return &csi.ControllerGetVolumeResponse{
+		Volume: &csi.Volume{
+			VolumeId: volume.Name,
+		},
+		Status: &csi.ControllerGetVolumeResponse_VolumeStatus{
+			PublishedNodeIds: exportList,
+			VolumeCondition: &csi.VolumeCondition{
+				Abnormal: abnormal,
+				Message:  message,
+			},
+		},
+	}, nil
+}
+
+func (s *service) getVolumeCondition(ctx context.Context, isiConfig *IsilonClusterConfig) (bool, string) {
```

Comment:
can we renamed this method to something like isVolumeAbnormal, as when we are returning false, we are saying volume is healthy.